### PR TITLE
feat(transform): Synchronize dataset structure and dataframe columns

### PR DESCRIPTION
**Pull request by dustmop**

### CODE REVIEW

**Comment by dustmop** [comment]
Fixes https://github.com/qri-io/starlib/issues/111.

File: docs/go.sum [hunk 1]
```diff
@@ -1363,6 +1363,10 @@ github.com/qri-io/qfs v0.6.1-0.20210809192005-052457575e43 h1:3Nlkdpj+MYaiAq7MmZ
 github.com/qri-io/qfs v0.6.1-0.20210809192005-052457575e43/go.mod h1:PCkSctLfc6OAquG5QtiUzddDkUE4wezDIvaUOaeuwu0=
 github.com/qri-io/starlib v0.5.1-0.20210922011334-69258ca1f1b7 h1:oKCnOxrnjrXrjcPcQnVb1RZAGqt9ftkQIYAAiDRyebg=
 github.com/qri-io/starlib v0.5.1-0.20210922011334-69258ca1f1b7/go.mod h1:NQ7kgFSsqf4TRuNbRD2ZQlBjQpH77ctFx0HpapFthQI=
+github.com/qri-io/starlib v0.5.1-0.20210923222026-17beaf4e5499 h1:eVsn9SFgsysCkfk2Hk06i4z7q/eLkUoHRVAwzP6+rJw=
+github.com/qri-io/starlib v0.5.1-0.20210923222026-17beaf4e5499/go.mod h1:NQ7kgFSsqf4TRuNbRD2ZQlBjQpH77ctFx0HpapFthQI=
+github.com/qri-io/starlib v0.5.1-0.20210923223108-83b84a6f961f h1:EK/jFRM0aULeBk4AjgA/ttBywoPRy8h+mq/ivHaJh8Q=
+github.com/qri-io/starlib v0.5.1-0.20210923223108-83b84a6f961f/go.mod h1:NQ7kgFSsqf4TRuNbRD2ZQlBjQpH77ctFx0HpapFthQI=
```
Comment (b5) [inline docs/go.sum:1369]: `go mod tidy`

File: go.sum [hunk 2]
```diff
@@ -1344,6 +1344,10 @@ github.com/qri-io/qfs v0.6.1-0.20210809192005-052457575e43 h1:3Nlkdpj+MYaiAq7MmZ
 github.com/qri-io/qfs v0.6.1-0.20210809192005-052457575e43/go.mod h1:PCkSctLfc6OAquG5QtiUzddDkUE4wezDIvaUOaeuwu0=
 github.com/qri-io/starlib v0.5.1-0.20210922011334-69258ca1f1b7 h1:oKCnOxrnjrXrjcPcQnVb1RZAGqt9ftkQIYAAiDRyebg=
 github.com/qri-io/starlib v0.5.1-0.20210922011334-69258ca1f1b7/go.mod h1:NQ7kgFSsqf4TRuNbRD2ZQlBjQpH77ctFx0HpapFthQI=
+github.com/qri-io/starlib v0.5.1-0.20210923222026-17beaf4e5499 h1:eVsn9SFgsysCkfk2Hk06i4z7q/eLkUoHRVAwzP6+rJw=
+github.com/qri-io/starlib v0.5.1-0.20210923222026-17beaf4e5499/go.mod h1:NQ7kgFSsqf4TRuNbRD2ZQlBjQpH77ctFx0HpapFthQI=
+github.com/qri-io/starlib v0.5.1-0.20210923223108-83b84a6f961f h1:EK/jFRM0aULeBk4AjgA/ttBywoPRy8h+mq/ivHaJh8Q=
+github.com/qri-io/starlib v0.5.1-0.20210923223108-83b84a6f961f/go.mod h1:NQ7kgFSsqf4TRuNbRD2ZQlBjQpH77ctFx0HpapFthQI=
```
Comment (b5) [inline go.sum:1350]: `go mod tidy`

File: transform/startf/ds/dataset.go [hunk 3]
```diff
@@ -2,24 +2,30 @@
...
 	"github.com/qri-io/qfs"
 	"github.com/qri-io/qri/base"
+	"github.com/qri-io/qri/dsref"
 	"github.com/qri-io/starlib/dataframe"
 	"github.com/qri-io/starlib/util"
 	"go.starlark.net/starlark"
 	"go.starlark.net/starlarkstruct"
 )
 
+var log = golog.Logger("ds")
```
Comment (b5) [inline transform/startf/ds/dataset.go:27]: `ds` feels vague. How about `stards` instead?

File: transform/startf/transform.go [hunk 4]
```diff
@@ -321,16 +320,10 @@ func (r *StepRunner) onCommit(ds *stards.Dataset) error {
...
-		return err
-	}
 
-	if _, ok := r.changeSet["body"]; !ok {
-		// the body has not changed, but the user still expects to see
-		// the number of entries the body has
-		if err := r.loadAndAssignPreviousStructureEntries(context.TODO(), ds); err != nil {
-			return err
-		}
+	ctx := context.TODO()
```
Comment (b5) [inline transform/startf/transform.go:324]: hmmm.... getting the right context into here is going to be tough

File: transform/startf/ds/dataset.go [hunk 5]
```diff
@@ -418,5 +464,95 @@ func (d *Dataset) AssignPreviousStructureEntries(entries int) {
...
+	}
+
+	newSchema := map[string]interface{}{
+		"type": "array",
+		"items": map[string]interface{}{
+			"type":  "array",
+			"items": cols,
+		},
+	}
+	d.ds.Structure.Schema = newSchema
```
Comment (b5) [inline transform/startf/ds/dataset.go:500]: need to defend against `Structure` being nil here. Same as in `AssignBodyFromDataframe`, but with no need for a shema b/c you're setting it:

```go
	st := d.ds.Structure
	if st == nil {
		st = &dataset.Structure{
			Format: "csv",
		}
	}
```

**Review by b5 — APPROVED** [review]

**Pull request merged**